notification_query.go

```diff
@@ -0,0 +1,248 @@
+package wmi
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/go-ole/go-ole"
+	"github.com/go-ole/go-ole/oleutil"
+	"github.com/hashicorp/go-multierror"
+	"github.com/scjalliance/comshim"
+)
+
+var (
+	// ErrAlreadyRunning is returned when NotificationQuery is already running.
+	ErrAlreadyRunning = errors.New("already running")
+)
+
+const (
+	wbemErrTimedOut            = 0x80043001
+	defaultNotificationTimeout = time.Second
+)
+
+// NotificationQuery represents subscription to the WMI events.
+// For more info see https://docs.microsoft.com/en-us/windows/desktop/wmisdk/swbemservices-execnotificationquery
+type NotificationQuery struct {
```

Comment:
Have you embedded decoder and mutex into the `NotificationQuery` on purpose?

Reply:
Embedded Decoder intended to provide more convenient interface to change unmarhsalling params:
```golang
query := NewNotificationQuery(ch, "SELECT .... ")
query.AllowMissingFields = true
```

Mutex was embedded just out of habbit. Could be easily replaced with a field if you think it's reasonable.

Reply:
Ok, I see. In regards to embedded mutex: I just can't find any case when users have to lock/unlock `NotificationQuery` from outside. So I guess it would be better to make the mutex private.

Reply:
How about runtime editing of .Decoder fields? :)

Reply:
Now I see its necessity :) Thanks for the explanation.